File: Leetcode/1-500/125 - Valid Palindrome.py
class Solution:
    def isPalindrome(self, s: str) -> bool:
        s=s.strip().replace(" ", "")
        s = ''.join(char for char in s if char.isalnum())
        s=s.lower()

        # two pointers
        for i in range(len(s)):
            if s[i] != s[len(s)-1-i]:
                return False

        return True

        #####

        # slicing
        """
        if s == s[::-1]:
            return True
        else:
            return False
        """

File: Leetcode/1-500/125 - Valid Palindrome.cpp
#include <cctype>
#include <string>
using namespace std;

// inefficient solution
/*
class Solution {
public:
  bool isPalindrome(string s) {
    string n = "";
    for (char i : s) {
      if (std::isalnum(i)) {
        //   cout << i;
        n += i;
      }
    }

    for (char &c : n) {
      c = std::tolower(c);
    }

    int l = 0;
    int r = n.size() - 1;

    while (l <= r) {
      if (n[l] != n[r]) {
        return false;
      }
      l++;
      r--;
    }
    return true;
  }
};
*/

// without creating new string
class Solution {
public:
  bool isPalindrome(string s) {
    int l = 0;
    int r = s.size() - 1;

    while (l < r) {
      while (l < r && !isalnum(s[l])) {
        l++;
      }

      while (l < r && !isalnum(s[r])) {
        r--;
      }

      if (tolower(s[l]) != tolower(s[r])) {
        return false;
      }

      l++;
      r--;
    }

    return true;
  }
};
